tools/mpremote/mpremote/pyboardextended.py

```diff
@@ -621,37 +621,66 @@ def mount_local(self, path):
         self.cmd = PyboardCommand(self.serial, fout, path)
         self.serial = SerialIntercept(self.serial, self.cmd)
 
-    def soft_reset_with_mount(self, out_callback):
+    def write_ctrl_d(self, out_callback):
         self.serial.write(b"\x04")
         if not self.mounted:
             return
 
-        # Clear flag while board reboots, it will be re-set once mounted.
-        self.mounted = False
+        # Read response from the device until it is quiet (with a timeout).
+        INITIAL_TIMEOUT = 0.5
+        QUIET_TIMEOUT = 0.2
+        FULL_TIMEOUT = 5
+        t_start = t_last_activity = time.monotonic()
+        data_all = b""
+        while True:
+            t = time.monotonic()
+            n = self.serial.inWaiting()
+            if n > 0:
+                data = self.serial.read(n)
+                out_callback(data)
+                data_all += data
+                t_last_activity = t
+            else:
+                if len(data_all) == 0:
+                    if t - t_start > INITIAL_TIMEOUT:
+                        return
```

Comment:
So does this mean that every time the board is in raw mode, or any other "special" mode where no immediate response is expected, ctrl-d will essentially introduce a 0.5s delay?

Actually I guess Raw mode is exited by ctrl-d so would normally return something, such as a prompt, so should be the 0.2s timeout.

Not saying that's necessarily a problem... Just wanting to check I'm following this right.

Reply:
You are right.  But it's no worse than the previous behaviour which was:
- if nothing mounted, return immediately (that's still the behaviour here)
- if something mounted then wait 0.5 seconds for at least one char
- if any chars received within that 0.5 seconds then assume the ctrl-D that was sent triggered a soft reset, and then wait for all chars to be received before remounting

So the improvements here are:
- don't ever wait longer than 5 seconds (FULL_TIMEOUT)
- always check for "MPY: soft reboot" and a valid prompt indicator (eg ">>> ") to indicate a soft reset was done